Firestore/core/src/firebase/firestore/api/write_batch.h

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright 2019 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_API_WRITE_BATCH_H_
+#define FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_API_WRITE_BATCH_H_
+
+#if !defined(__OBJC__)
+#error "This header only supports Objective-C++"
+#endif  // !defined(__OBJC__)
+
+#import <Foundation/Foundation.h>
+
+#include <algorithm>
+#include <memory>
+#include <utility>
+#include <vector>
+
+#import "Firestore/Source/Core/FSTFirestoreClient.h"
+#import "Firestore/Source/Model/FSTMutation.h"
+
+#include "Firestore/core/src/firebase/firestore/api/document_reference.h"
+#include "Firestore/core/src/firebase/firestore/api/firestore.h"
+#include "Firestore/core/src/firebase/firestore/api/input_validation.h"
+#include "Firestore/core/src/firebase/firestore/core/user_data.h"
+#include "Firestore/core/src/firebase/firestore/util/status.h"
+
+NS_ASSUME_NONNULL_BEGIN
+
+@class FSTMutation;
+
+namespace firebase {
+namespace firestore {
+namespace api {
+
+class WriteBatch {
+ public:
+  WriteBatch() = default;
```

Comment:
Gil added a `util::DelayedConstructor` class recently to avoid having classes declare default constructors when they don't have a meaningful empty state. If you feel like `WriteBatch` has a valid default state, leave as is; otherwise, you can eliminate this default constructor and store a `DelayedConstructor<WriteBatch>` instead of just `WriteBatch` in `FIRWriteBatch`, then use the `DelayedConstructor<>::init` method. Conceptually, `DelayedConstructor` is very similar to `absl::optional`.

Reply:
Ahh, right.  Done, thanks!